### PR TITLE
feat(ci): add binary compatiblity test

### PR DESCRIPTION
**Pull request by tsunrise**
## Summary
Add test to support binary compatibility of UTXO. 

## Details
This PR ports the hardcoded test from https://github.com/Manta-Network/Manta/blob/manta/pallets/manta-pay/src/bin/precompute_coins.rs. This test makes sure changes in current PR are still compatible with the UTXO versions in `manta-sdk`. 

Closes: #89 

---

Before we can merge this PR, please make sure that all the following items have been checked off:

- [x] Linked to an issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Added **one** line describing your change in [`CHANGELOG.md`](https://github.com/manta-network/manta-rs/blob/main/CHANGELOG.md) and added the appropriate `changelog` label to the PR.
- [x] Re-reviewed `Files changed` in the GitHub PR explorer.
- [x] Checked that changes and commits conform to the standards outlined in [`CONTRIBUTING.md`](https://github.com/manta-network/manta-rs/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: manta-pay/Cargo.toml [hunk 1]
```diff
@@ -140,4 +140,6 @@ ws_stream_wasm = { version = "0.7.3", optional = true, default-features = false
 [dev-dependencies]
 manta-crypto = { path = "../manta-crypto", default-features = false, features = ["getrandom"] }
 manta-pay = { path = ".", default-features = false, features = ["groth16", "scale", "scale-std", "std", "test"] }
+manta-sdk = { git = "https://github.com/manta-network/sdk.git", default-features = false, features = ["download"]}
```
Comment (stechu) [inline manta-pay/Cargo.toml:143]: This needs to anchor a specific `manta-sdk` version. (cc @bhgomes )

**Review by stechu — CHANGES_REQUESTED** [review]

File: manta-pay/Cargo.toml (hunk 1 above)
Comment (tsunrise) [reply manta-pay/Cargo.toml:143]: @bhgomes which `manta-sdk` version shall we use?

Comment (bhgomes) [reply manta-pay/Cargo.toml:143]: We're going to have to think about this. We might have to move the "concrete parameters" out of SDK. I'll meditate on this over the weekend.

File: manta-pay/src/test/compat.rs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! ## Manta Pay UTXO Binary Compatibility
+//! This test checks if current implementation is compatible with precomputed transactions in Manta SDK.
+
+// This test is adapted from https://github.com/Manta-Network/Manta/blob/1c5c4b8750d06cb028fb555813414a5802554817/pallets/manta-pay/src/bin/precompute_coins.rs
```
Comment (BoyuanFeng) [inline manta-pay/src/test/compat.rs:20]: Not sure if we should keep this comment.

File: manta-pay/src/test/compat.rs [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+        proving_context,
+        verifying_context,
+        parameters,
+        UtxoAccumulatorModel::decode(
+            manta_sdk::pay::testnet::parameters::UtxoAccumulatorModel::get()
+                .expect("Checksum did not match."),
+        )
+        .expect("Unable to decode UTXO_ACCUMULATOR_MODEL."),
+    ))
+}
```
Comment (BoyuanFeng) [inline manta-pay/src/test/compat.rs:115]: This function may be updated once [PR#102](https://github.com/Manta-Network/manta-rs/pull/102) is merged.

File: manta-pay/src/test/compat.rs [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+        ProofSystem::verify(
+            verifying_context,
+            &post.generate_proof_input(),
+            &post.validity_proof,
+        )
+        .expect("Unable to verify proof."),
+        "Invalid proof: {:?}.",
+        post
+    );
+}
```
Comment (BoyuanFeng) [inline manta-pay/src/test/compat.rs:130]: The same function is also used in [manta-benchmark](https://github.com/Manta-Network/manta-rs/blob/main/manta-benchmark/src/payment.rs#L38-L51). Should we keep only one?

File: manta-pay/src/test/compat.rs [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    )
+    .into_post(
+        FullParameters::new(parameters, utxo_accumulator.model()),
+        &proving_context.reclaim,
+        rng,
+    )
+    .expect("Unable to build RECLAIM proof.");
+    assert_valid_proof(&verifying_context.reclaim, &reclaim);
+    ([mint_0, mint_1], reclaim)
+}
```
Comment (BoyuanFeng) [inline manta-pay/src/test/compat.rs:268]: These functions are also used in [manta-benchmark](https://github.com/Manta-Network/manta-rs/blob/main/manta-benchmark/src/payment.rs#L53-L164). Maybe we only need one copy.

File: manta-pay/src/test/compat.rs [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    let (proving_context, verifying_context, parameters, utxo_accumulator_model) =
+        load_parameters(directory.path()).expect("failed to load parameters");
+    let asset_id: u32 = 8;
+
+    let _ = sample_mint(
+        &proving_context.mint,
+        &verifying_context.mint,
+        &parameters,
+        &utxo_accumulator_model,
+        AssetId(asset_id).value(100_000),
```
Comment (BoyuanFeng) [inline manta-pay/src/test/compat.rs:286]: may use `rng.gen()`

File: manta-pay/src/test/compat.rs [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+/// Test validity on sampled transactions.
+#[test]
+fn compatibility() {
+    let directory = tempfile::tempdir().expect("Unable to generate temporary test directory.");
+    println!("[INFO] Temporary Directory: {:?}", directory);
+
+    let mut rng = thread_rng();
+    let (proving_context, verifying_context, parameters, utxo_accumulator_model) =
+        load_parameters(directory.path()).expect("failed to load parameters");
+    let asset_id: u32 = 8;
```
Comment (BoyuanFeng) [inline manta-pay/src/test/compat.rs:279]: delete this line

File: manta-pay/src/test/compat.rs [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+        AssetId(asset_id).value(100_000),
+        &mut rng,
+    );
+    let _ = sample_private_transfer(
+        &proving_context,
+        &verifying_context,
+        &parameters,
+        &utxo_accumulator_model,
+        AssetId(asset_id).value(10_000),
+        AssetId(asset_id).value(20_000),
```
Comment (BoyuanFeng) [inline manta-pay/src/test/compat.rs:295]: may use `rng.gen()`

File: manta-pay/src/test/compat.rs [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+        &mut rng,
+    );
+    let _ = sample_reclaim(
+        &proving_context,
+        &verifying_context,
+        &parameters,
+        &utxo_accumulator_model,
+        AssetId(asset_id).value(10_000),
+        AssetId(asset_id).value(20_000),
```
Comment (BoyuanFeng) [inline manta-pay/src/test/compat.rs:304]: may use `rng.gen()`

File: manta-pay/src/test/mod.rs [hunk 10]
```diff
@@ -24,3 +24,5 @@
 // pub mod simulation;
 
 pub mod transfer;
+
+pub mod compat;
```
Comment (BoyuanFeng) [inline manta-pay/src/test/mod.rs:28]: Should we change the order here?
`pub mod compat;`
`pub mod transfer;`

**Review by BoyuanFeng — CHANGES_REQUESTED** [review]

File: manta-pay/src/test/mod.rs [hunk 11]
```diff
@@ -23,4 +23,5 @@
 // #[cfg_attr(doc_cfg, doc(cfg(feature = "simulation")))]
 // pub mod simulation;
 
+pub mod compat;
```
Comment (bhgomes) [inline manta-pay/src/test/mod.rs:26]: Let's use full names `compatibility`.

File: manta-pay/src/lib.rs [hunk 12]
```diff
@@ -41,6 +41,10 @@ pub mod key;
 #[cfg_attr(doc_cfg, doc(cfg(all(feature = "groth16", feature = "test"))))]
 pub mod parameters;
 
+#[cfg(all(feature = "groth16", feature = "test"))]
+#[cfg_attr(doc_cfg, doc(cfg(all(feature = "groth16", feature = "test"))))]
+pub mod payment;
```
Comment (bhgomes) [inline manta-pay/src/lib.rs:46]: Can we move this module into `test`? I think the name is very misleading at the top level.

Comment (bhgomes) [reply manta-pay/src/lib.rs:46]: Also `assert_valid_proof` should at least be moved into `manta_accounting::transfer::test`.

**Review by bhgomes — CHANGES_REQUESTED** [review]

File: manta-pay/src/test/compat.rs [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    Parameters,
+    UtxoAccumulatorModel,
+)> {
+    println!("Loading parameters...");
+    let mint_path = directory.join("mint.dat");
+    manta_parameters::pay::testnet::proving::Mint::download(&mint_path)?;
+    let private_transfer_path = directory.join("private-transfer.dat");
+    manta_parameters::pay::testnet::proving::PrivateTransfer::download(&private_transfer_path)?;
+    let reclaim_path = directory.join("reclaim.dat");
+    manta_parameters::pay::testnet::proving::Reclaim::download(&reclaim_path)?;
```
Comment (bhgomes) [inline manta-pay/src/test/compat.rs:50]: Try to avoid `download` here. We have the files locally, so we should be able to read them directly.

Comment (BoyuanFeng) [reply manta-pay/src/test/compat.rs:50]: This `download` is only for lfs files in `ProvingVontext`. A developer does NOT have files locally unless he has intentionally run `git lfs pull` earlier.

File: manta-pay/src/lib.rs (hunk 12 above)
Comment (BoyuanFeng) [reply manta-pay/src/lib.rs:46]: > Can we move this module into `test`? I think the name is very misleading at the top level.

After moving into `test`, we cannot import this module in `manta-benchmark`. Should we rename it as `sample_payment.rs`?

Comment (BoyuanFeng) [reply manta-pay/src/lib.rs:46]: > Also `assert_valid_proof` should at least be moved into `manta_accounting::transfer::test`.

`assert_valid_proof` requires `config` in `manta-pay`. Moving this function to `manta-accounting` will introduce cycle dependency.

Comment (bhgomes) [reply manta-pay/src/lib.rs]: No, generalize it to using `transfer::Configuration`.

File: manta-pay/src/lib.rs [hunk 14]
```diff
@@ -23,8 +23,9 @@
 
 extern crate alloc;
 
-#[cfg(test)]
-mod test;
+#[cfg(feature = "test")]
```
Comment (BoyuanFeng) [inline manta-pay/src/lib.rs:26]:
```suggestion
#[cfg(any(test, feature = "test"))]
```

File: manta-pay/src/lib.rs [hunk 15]
```diff
@@ -23,8 +23,9 @@
...
+#[cfg_attr(doc_cfg, doc(cfg(feature = "test")))]
+pub mod test;
```
Comment (BoyuanFeng) [inline manta-pay/src/lib.rs:28]: Move into sorted order with other modules.

File: manta-pay/Cargo.toml [hunk 16]
```diff
@@ -1,17 +1,17 @@
 [package]
```
Comment (BoyuanFeng) [inline manta-pay/Cargo.toml:1]: use old form of Cargo.toml

File: manta-pay/src/test/compatibility.rs [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+        )
+        .expect("Unable to decode UTXO_ACCUMULATOR_MODEL."),
+    ))
+}
+
+/// Test validity on sampled transactions.
+#[test]
+fn compatibility() {
+    let directory = tempfile::tempdir().expect("Unable to generate temporary test directory.");
+    let mut rng = thread_rng();
```
Comment (BoyuanFeng) [inline manta-pay/src/test/compatibility.rs:112]: use `OsRng`

File: manta-pay/src/test/compatibility.rs [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+        parameters,
+        UtxoAccumulatorModel::decode(
+            manta_parameters::pay::testnet::parameters::UtxoAccumulatorModel::get()
+                .expect("Checksum did not match."),
+        )
+        .expect("Unable to decode UTXO_ACCUMULATOR_MODEL."),
+    ))
+}
+
+/// Test validity on sampled transactions.
```
Comment (BoyuanFeng) [inline manta-pay/src/test/compatibility.rs:108]: Update docs: "Test that the circuit has not changed against the current known parameters in `manta-parameters`" (something like this)

File: manta-pay/src/test/compatibility.rs [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+use crate::{
+    config::{
+        MultiProvingContext, MultiVerifyingContext, NoteEncryptionScheme, Parameters,
+        ProvingContext, UtxoAccumulatorModel, UtxoCommitmentScheme, VerifyingContext,
+        VoidNumberCommitmentScheme,
+    },
+    test::payment::{prove_mint, prove_private_transfer, prove_reclaim},
+};
+use anyhow::Result;
+use ark_std::rand::thread_rng;
```
Comment (BoyuanFeng) [inline manta-pay/src/test/compatibility.rs:29]: never ever use ark_std

**Review by bhgomes — APPROVED** [review]

**Pull request merged**